### PR TITLE
Add 2025 English IMD data with year-based API and comprehensive test suite

### LINKED ISSUE #19: tests
There are currently no tests. 
- [x] add pytest-django to requirements
- [x] curate a list of gold standard IMDs against postcodes for England
- [x] curate a list of gold standard IMDs against postcodes for Wales
- [x] curate a list of gold standard IMDs against postcodes for Scotland
- [x] curate a list of gold standard IMDs against postcodes for Northern Ireland
- [x] test script to compare gold standard IMD raw score, rank against API
- [x] test script to compare quantile against gold standard rank

### LINKED ISSUE #34: 2025 deprivation scores
https://www.gov.uk/government/statistics/english-indices-of-deprivation-2025/english-indices-of-deprivation-2025-statistical-release

https://www.theguardian.com/society/ng-interactive/2025/oct/30/how-deprived-is-your-area

We need to integrate this new data and and retain a way of querying the previous data. The audit platforms should use the old data for previous audit years.

### PR DESCRIPTION
**Pull request by eatyourpeas**
## Summary

This PR adds support for the newly released **2025 English Index of Multiple Deprivation (IMD)** data, alongside significant improvements to the API and a comprehensive test suite.

The 2025 IMD data published 30/10/2025 follows the same format as previous publications, but uses the 2021 LSOAs rather than the 2011 ones. In practice these are the same but in 2021 there are more of them (33,755 vs 32844 in 2011). A vulnerability is that `postcodes.io` which this uses returns still the 2011 list - note [this issue](https://github.com/ideal-postcodes/postcodes.io/issues/1181) open since 2023. I am not clear what LSOA is returned for one of the new ones. It may be that if in time they return 2021 LSOA codes we shall need to update. Because there are now multiple datasets for the same thing, there are new migrations as the models now have a year field which is `unique_together` with the code.

In addition to this, the api now accepts a year as an optional parameter. There are graceful fall backs if none is provided, defaulting to 2019 for England and Wales, 2017 for Northern Ireland and 2020 for Scotland. 

There are now also tests for all of these datasets

closes #19 
closes #34  

#### 🆕 2025 IMD Data Support
- Added 2025 English IMD data using 2021 LSOAs and 2024 Local Authority Districts
- Models updated to include `year` field with `unique_together` constraints
- API now accepts optional `year` parameter (defaults to 2019 for backward compatibility)
- Quantile functions updated to support year-based calculations

#### 🧪 Comprehensive Test Suite (125 tests)
- **31 English tests** - covering 2019 and 2025 IMD data
- **18 Scottish tests** - SIMD 2020 data
- **37 Welsh tests** - WIMD 2019 data  
- **39 Northern Ireland tests** - NIMDM 2017 data
- Tests cover IMD rank calculations, quantile calculations, edge cases (wrong year, missing quantile, invalid postcodes)

#### 🔧 CI/CD Infrastructure
- New GitHub Actions workflow (`test.yml`) runs tests on PRs to `live` and `development`
- Mocked postcodes.io API responses for reliable CI testing (62 postcodes captured)
- New `ci_test` seed mode for faster database seeding (~3.5 min total CI time)
- Pytest-django integration with existing seeded database

#### 📝 Additional Improvements
- Guards added to skip seeding when data already exists
- Fixed quantile handling for impossible quantile values
- Removed tqdm progress bar (didn't work well in Docker)
- Added Northern Ireland postcode scraping script for test data generation

### Commits (9)
1. `edc995e` - 2025 England IMD and 2021 LSOAs, 2024 LADs - added year to models
2. `6cb5e22` - Table totals complete with documentation
3. `b2b1b66` - Remove progress bar/tqdm
4. `229235d` - Guards to skip when data already seeded
5. `50dcd87` - Refactor IMD endpoints to accept year
6. `bcaad1d` - Implement pytest, year parameter, tests for 2019/2025
7. `d6eb78f` - 31 English, 18 Scottish, 37 Welsh tests + quantile fixes
8. `64bcd64` - NI postcode scraping script, 39 NI tests
9. `c4e6feb` - Mocking postcodes.io, CI workflow, ci_test seed mode

### Testing
All 125 tests pass with mocked postcodes.io responses:
```
125 passed, 125 warnings in 5.09s
```

### Breaking Changes
None - the `year` parameter is optional and defaults to 2019 for backward compatibility.

### CODE REVIEW

**Comment by eatyourpeas** [comment]
@mbarton requesting you to run this up before merging if you have time as it feels a reasonably big new thing and although not a breaking change, it has touched what was working before - what I have not been able to test is the custom migration that updates existing records with the correct year. If you have a version of live cloned and then run this PR I would be grateful as this is one thing I have not done.

File: tests/conftest.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    """
+    Mock postcodes.io API calls in tests.
+
+    By default, mocking is DISABLED to allow integration tests to hit the real API.
+
+    To enable mocking:
+    - Set environment variable MOCK_POSTCODES_IO=1
+    - Or use the @pytest.mark.mock_postcodes_io marker on individual tests
+
+    This allows CI to run tests against the real postcodes.io API while
```
Comment (mbarton) [inline tests/conftest.py:70]: I think we should use the mock in all cases - we don't want to be blocked on merging a change to this repo because a (free!) third party service is having issues

**Review by mbarton — APPROVED** [review]
I've run up `live` locally, seeded then checked out this PR, run the migrations and seeded the new data. The `index_of_multiple_deprivation_quantile` endpoint seems to work fine after.

I've got a small comment about the tests but happy to deal with that in an additional PR since this one is already quite big

**Comment by eatyourpeas** [comment]
I am pleased to say that the RCPCH `postcodes.io` instance does return both 2011 and 2021 codes for LSOAs. This is an example of an LSOA that has a different code in 2011 to 2021.
```
{
  "status": 200,
  "result": {
    "postcode": "TS18 3WH",
    "quality": 1,
    "eastings": 443620,
    "northings": 517339,
    "country": "England",
    "nhs_ha": "North East",
    "longitude": -1.327179,
    "latitude": 54.549443,
    "european_electoral_region": "North East",
    "primary_care_trust": "Stockton-on-Tees Teaching",
    "region": "North East",
    "lsoa": "Stockton-on-Tees 026C",
    "msoa": "Stockton-on-Tees 026",
    "incode": "3WH",
    "outcode": "TS18",
    "parliamentary_constituency": "Stockton North",
    "parliamentary_constituency_2024": "Stockton North",
    "admin_district": "Stockton-on-Tees",
    "parish": "Stockton-on-Tees, unparished area",
    "admin_county": null,
    "date_of_introduction": "201711",
    "admin_ward": "Ropner",
    "ced": null,
    "ccg": "NHS North East and North Cumbria",
    "nuts": "Stockton-on-Tees",
    "pfa": "Cleveland",
    "nhs_region": "North East and Yorkshire",
    "ttwa": "Middlesbrough and Stockton",
    "national_park": "England (non-National Park)",
    "bua": "Stockton-on-Tees",
    "icb": "NHS North East and North Cumbria Integrated Care Board",
    "cancer_alliance": "Northern",
    "lsoa11": "Stockton-on-Tees 014C",
    "msoa11": "Stockton-on-Tees 014",
    "lsoa21": "Stockton-on-Tees 026C",
    "msoa21": "Stockton-on-Tees 026",
    "oa21": "E00188233",
    "ruc11": "(England/Wales) Urban city and town",
    "ruc21": "Urban: Nearer to a major town or city",
    "lep1": "Tees Valley",
    "lep2": null,
    "codes": {
      "admin_district": "E06000004",
      "admin_county": "E99999999",
      "admin_ward": "E05014878",
      "parish": "E43000004",
      "parliamentary_constituency": "E14001518",
      "parliamentary_constituency_2024": "E14001518",
      "ccg": "E38000247",
      "ccg_id": "16C",
      "ced": "E99999999",
      "nuts": "TLC31",
      "lsoa": "E01035201",
      "msoa": "E02007061",
      "lau2": "E06000004",
      "pfa": "E23000013",
      "nhs_region": "E40000012",
      "ttwa": "E30000093",
      "national_park": "E65000001",
      "bua": "E63007405",
      "icb": "E54000050",
      "cancer_alliance": "E56000029",
      "lsoa11": "E01012264",
      "msoa11": "E02002548",
      "lsoa21": "E01035201",
      "msoa21": "E02007061",
      "oa21": "E00188233",
      "ruc11": "C1",
      "ruc21": "UN1",
      "lep1": "E37000034",
      "lep2": null
    }
  }
}
```

EDIT - the live `api.postcodes.io` service has this updated object, but the RCPCH does not

File: tests/conftest.py (hunk 1 above)
Comment (eatyourpeas) [reply tests/conftest.py:70]: Yes I agree. I have changed the flag ro `REAL_POSTCODES_IO` and by default it is false, so mocking occurs. If for some reason, maybe in dev, we want to use the real api we set the flag as we run pytest.
